intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp

```diff
@@ -0,0 +1,317 @@
+  last_n_tokens.resize(n_ctx, 0);
+  token_eos = false;
+  curr_input_ids.clear();
+}
+
+std::string Model::generate_one_token(const std::string& prompt) {
+  if (curr_input_ids.empty()) {
+    auto embd_inp = ::model_tokenize(ctx, prompt, false);
+    curr_input_ids = embd_inp;
+  }
+  for (auto item : curr_input_ids) {
+    last_n_tokens.erase(last_n_tokens.begin());
+    last_n_tokens.push_back(item);
+  }
+  model_eval(ctx, &curr_input_ids[0], curr_input_ids.size(), n_past, params.n_threads);
+  n_past += curr_input_ids.size();
+
+  float* logits = model_get_logits(ctx);
+  int next_token_id = post_process(logits);
+  curr_input_ids = {next_token_id};
+
+  if (next_token_id == model_token_eos() || n_past - prompt.size() == params.n_predict) {
+    token_eos = true;
+  }
+
+  auto next_token = model_token_to_str(ctx, next_token_id);
+  if (strcmp(next_token, "<|endoftext|>") == 0) {
+    token_eos = true;
+  }
+
+  return next_token;
+}
+
+std::string Model::generate_tokens(const std::string& prompt) {
+  int n_past = 0;
+  int n_remain = params.n_predict;
+  int max_length = 512;
+  auto embd_inp = ::model_tokenize(ctx, prompt, false);
+  int n_eval = embd_inp.size();
+  std::vector<int> curr_input_ids(embd_inp);
+  std::vector<int> output_ids;
+  output_ids.reserve(max_length);
+  std::string ret;
+  ret += prompt;
+  while (output_ids.size() < n_remain) {
+    for (auto item : curr_input_ids) {
+      last_n_tokens.erase(last_n_tokens.begin());
+      last_n_tokens.push_back(item);
+    }
+    model_eval(ctx, &curr_input_ids[0], curr_input_ids.size(), n_past, params.n_threads);
+    n_past += curr_input_ids.size();
+
+    float* logits = model_get_logits(ctx);
+    int next_token_id = post_process(logits);
+    curr_input_ids = {next_token_id};
+
+    output_ids.push_back(next_token_id);
+    ret += model_token_to_str(ctx, next_token_id);
+
+    if (next_token_id == model_token_eos()) {
+      break;
+    }
+  }
+
+  return ret;
+}
+
+std::string Model::generate(const std::string& prompt, bool stream_mode) {
+  if (stream_mode) {
+    return generate_one_token(prompt);
+  }
+
+  return generate_tokens(prompt);
+}
+
+int Model::post_process(float* logits) {
+  int alpha_frequency = 0;
+  int alpha_presence = 0;
+  int repeat_last_n = 64;
+  int top_k = 40;
+  float tfs_z = 1.00f;
+  float typical_p = 1.00f;
+  float top_p = 0.95f;
+  float temp = 0.80f;
+  std::vector<model_token_data> candidates;
+  candidates.reserve(n_vocab);
+  for (model_token token_id = 0; token_id < n_vocab; token_id++) {
+    candidates.emplace_back(model_token_data{token_id, logits[token_id], 0.0f});
+  }
+  model_token_data_array candidates_p = {candidates.data(), candidates.size(), false};
+
+  // Apply penalties
+  float nl_logit = logits[model_token_nl()];
+  auto last_n_repeat = std::min(std::min((int)last_n_tokens.size(), repeat_last_n), n_ctx);
+  model_sample_repetition_penalty(ctx, &candidates_p, last_n_tokens.data() + last_n_tokens.size() - last_n_repeat,
```

Comment:
chatbot only need this kind of post processing?

Reply:
More post processing will be add, maybe another pr?

Reply:
how about leaving an option for post processing in advance? In this PR we only support ```--post-process (default)greedy-search```

Reply:
will be fix in this pr

Reply:
arguments added, function will be refined.